tests/datasim.py

```diff
@@ -0,0 +1,833 @@
+            is applied to ``value_to_fix`` argument.
+        """
+        self._values_to_fix[index] = value_to_fix
+        self._values_after_fix[index] = value_after_fix
+
+    def update_column_values(
+        self, values_to_fix: Sequence[Any], values_after_fix: Sequence[Any]
+    ):
+        """
+        Update a single column sample with new values
+
+        Parameters
+        ----------
+        values_to_fix : Sequence[Any]
+            Raw values that need a fix (usually values containing errors that have
+            been simulated by a ReverseFeatureOperation).
+        values_to_fix : Sequence[Any]
+            Values that are supposed to be found after that the proper correction
+            is applied to ``values_to_fix`` argument.
+        """
+        self._values_to_fix = values_to_fix
+        self._values_after_fix = values_after_fix
+
+
+class TestDataSet:
+    def __init__(self, sample_size: Optional[int] = None):
+        # Creating two dicts with the same elements, so that
+        # retrieving an element by name or index takes the same time
+        self._columns_by_index = []
+        self._sample_size = sample_size
+        self._name_to_index_map = {}
+
+    @property
+    def sample_size(self) -> Optional[int]:
+        return self._sample_size
+
+    def _validate_testcolumn_to_create(self, column: TestColumn):
+        """
+        Check the TestColumn instance ``column`` before inserting it in the dataset
+
+        This method checks that:
+        - ``column`` is an instance of TestColumn
+        - no column with the same name is present in the dataset
+        - the ``column`` has the same number of samples as the others
+
+        Parameters
+        ----------
+        column : TestColumn
+            Column that needs to be validated.
+
+        Raises
+        ------
+        TypeError
+            If the ``column`` argument is not a TestColumn instance.
+        ValueError
+            If another column with the same name is already present in the
+            instance, or if the column has a different sample count from the
+            other instance columns.
+        """
+        if not isinstance(column, TestColumn):
+            raise TypeError(
+                "The 'column' argument must be a TestColumn instance, instead"
+                + f" its type is {type(column)}"
+            )
+        if column.name in self._name_to_index_map:
+            raise ValueError(
+                "The name and/or the column id is already present in the "
+                "Dataset. If an update is required, use __setitem__ by indexing"
+            )
+        if len(column) != self.sample_size:
+            raise ValueError(
+                f'`column` argument named "{column.name}" has {len(column)}'
+                f" values, while the other dataset columns have {self.sample_size}"
+                " values."
+            )
+
+    def _validate_testcolumn_to_update(self, column: _TestColumn):
```

Comment:
I would have preferred to have one function, now 2/3 of the code inside is duplicated...